ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -377,6 +426,79 @@ extendVRNew cfg tx
+-- possible, returning the last 'ValidationResult' and the remaining
+-- transactions which couldn't be added due to the mempool capacity being
+-- reached.
+validateNew :: forall m blk. (MonadSTM m, ApplyTx blk)
+            => MempoolEnv m blk
+            -> [GenTx blk]
+            -> ValidationResult blk
+            -> STM m (ValidationResult blk, [GenTx blk])
+validateNew mpEnv txs res =
+  let res' = res { vrInvalid = [] }
+  in case txs of
+    []                  -> return (res', [])
+    headTx:remainingTxs -> do
+      -- First, attempt to individually validate the first new transaction.
+      --
+      -- If this fails due to the mempool capacity being met, we should simply
+      -- 'retry' as it will be cheap due to the fact that we've done very
+      -- little work in this STM transaction.
+      --
+      -- If, however, this is successful, we should continue to validate all
+      -- of the other new transactions one-by-one. If the mempool capacity
+      -- would be reached at any step, we update the 'InternalState' with the
+      -- work that we've already done and return the last 'ValidationResult'
+      -- along with the remaining unvalidated transactions.
+      --
+      -- @intricate: I'm actually not sure if it makes sense to do this
+      --             specific validation of the head tx. 'go' looks like it's
+      --             sufficient for handling all of the transactions. If 'go'
+      --             fails on the first transaction (due to mempool cap), it
+      --             will return the remaining unvalidated transactions all the
+      --             way back up to 'implAddTxs' and, as a result, 'implAddTxs'
+      --             will recurse.
+      headTxValidationRes <- go [headTx] res'
+      case headTxValidationRes of
+        -- Mempool capacity hasn't been reached (no remaining unvalidated
+        -- transactions were returned).
+        (vr, []) -> do
+          -- Continue validating the remaining transactions.
+          (vr', unvalidatedTxs) <- go remainingTxs vr
+          writeTVar mpEnvStateVar IS { isTxs          = vrValid        vr'
+                                     , isTip          = vrBefore       vr'
+                                     , isLastTicketNo = vrLastTicketNo vr'
+                                     }
+          pure (vr', unvalidatedTxs)
+
+        -- The mempool capacity has been reached.
+        _ -> retry
+ where
+  MempoolEnv
+    { mpEnvStateVar
+    , mpEnvLedgerCfg
+    , mpEnvCapacity
+    } = mpEnv
+
+  go :: [GenTx blk]
+     -> ValidationResult blk
+     -> STM m (ValidationResult blk, [GenTx blk])
+  go []      vr = pure (vr, [])
+  go (t:ts)  vr = do
+    mempoolSize <- getMempoolSize mpEnv
+    let MempoolCapacity mempoolCap = mpEnvCapacity
+    -- The size of a mempool should never be greater than its capacity.
+    assert (mempoolSize <= mempoolCap) $
+      -- @intricate: Should we first validate the transaction and then check
```

Comment:
@mrBliss: curious about your opinion here. I'm not sure it's worth the extra work to validate a transaction in hopes that it's invalid when we already know that we're at the mempool's capacity.

Reply:
Yeah, as am I (also leaning towards not validating first). This shouldn't be hard to change anyways. I'll document what's going on in this part of the code so at least we'll know exactly where to change it in the future, if necessary.